frontend/packages/dev-console/src/components/pipelineruns/logs/logs-utils.ts

```diff
@@ -0,0 +1,142 @@
+  );
+  return {
+    containers: containers.slice(
+      0,
+      firstRunningCont === -1 ? containers.length : firstRunningCont + 1,
+    ),
+    stillFetching: firstRunningCont !== -1,
+  };
+};
+
+const getOrderedStepsFromPod = (name: string, ns: string): Promise<ContainerStatus[]> => {
+  return k8sGet(PodModel, name, ns)
+    .then((pod: PodKind) => {
+      return getSortedContainerStatus(
+        pod.spec.containers ?? [],
+        pod.status?.containerStatuses ?? [],
+      );
+    })
+    .catch((err) => {
+      errorModal({ error: err.message || 'Error downloading logs.' });
+      return [];
+    });
+};
+
+type StepsWatchUrl = {
+  [key: string]: {
+    name: string;
+    steps: { [step: string]: string };
+  };
+};
+
+export const getDownloadAllLogsCallback = (
+  sortedTaskRuns: string[],
+  taskRunFromYaml: TaskRuns,
+  namespace: string,
+  pipelineRunName: string,
+): (() => void) => {
+  const getWatchUrls = (): Promise<StepsWatchUrl> =>
+    sortedTaskRuns.reduce(async (acc, currTask) => {
+      const { pipelineTaskName, status = {} } = taskRunFromYaml[currTask];
+      const { podName } = status;
+      const steps = await getOrderedStepsFromPod(podName, namespace);
+      const allStepUrls = steps.reduce((stepUrls, currentStep) => {
+        const { name } = currentStep;
+        if (containerToLogSourceStatus(currentStep) !== LOG_SOURCE_TERMINATED) return stepUrls;
+        const urlOpts = {
+          ns: namespace,
+          name: podName,
+          path: 'log',
+          queryParams: {
+            container: name,
+            follow: 'true',
+          },
+        };
+        return { ...stepUrls, [name]: resourceURL(PodModel, urlOpts) };
+      }, {});
+      const accumulator = await acc;
+      accumulator[currTask] = {
+        name: pipelineTaskName,
+        steps: { ...allStepUrls },
+      };
+      return Promise.resolve(accumulator);
+    }, Promise.resolve({}));
+
+  const fetchLogs = async (tasksPromise: Promise<StepsWatchUrl>) => {
+    const tasks = await tasksPromise;
+    const { headings, urls: allRequests } = sortedTaskRuns.reduce(
+      (acc, currTask) => {
+        const task = tasks[currTask];
+        const urls: Promise<string>[] = Object.keys(task.steps).map((step) =>
+          coFetchText(task.steps[step]),
+        );
+        const newHeadings: string[] = Object.keys(task.steps).map((step, i) => {
+          if (i === 0) return `${task.name}\n\n${step}\n\n`;
+          return `${step}\n\n`;
+        });
+        return {
+          headings: [...acc.headings, ...newHeadings],
+          urls: [...acc.urls, ...urls],
+        };
+      },
+      { headings: [], urls: [] },
+    );
+    const buffer = new LineBuffer(null);
+    return Promise.all(allRequests)
+      .then((allLogs) => {
+        allLogs.forEach((logs, i) => {
+          buffer.ingest(headings[i]);
+          buffer.ingest(`${logs}\n\n`);
+        });
```

Comment:
I think we can do better for separation of concerns and allowing for cancellations if it's waiting forever to complete (because we want running logs too - but they can stall for number of reasons).

```suggestion
    const allRequests: Promise<string>[] = sortedTaskRuns.reduce((acc, currTask) => {
      const task = tasks[currTask];
      const promises: Promise<string>[] = Object.keys(task.steps).map((step, i) => {
        let heading = '';
        if (i === 0) {
          heading += `${task.name}\n\n`;
        }
        heading += `${step}\n\n`;

        const { url, status } = task.steps[step];
        const getContentPromise = coFetchText(url).then((logs) => {
          return `${heading}${logs}\n\n`;
        });

        if (status === LOG_SOURCE_TERMINATED) {
          // If we are done, we want this log content
          return getContentPromise;
        }

        // If we are not done, let's not wait indefinitely
        return Promise.race([
          getContentPromise,
          new Promise<string>((resolve) => {
            setTimeout(() => resolve(''), 1000);
          }),
        ]);
      },
      );
      return [...acc, ...promises];
    }, []);

    const buffer = new LineBuffer(null);
    return Promise.all(allRequests).then((stepLogs) => {
      buffer.ingest(stepLogs.join(''));
```